File: solutions/javascript/blind-75/string/longestSubstringWithoutRepeatingCharacters.js
const lengthOfLongestSubstring = (string) => {
    let maxValidSubstringLength = 0;
    const charsInSubstring = new Set();
    
    let windowStart = 0;
    let windowEnd = 0;
    
    const numChars = string.length;
    while (windowEnd < numChars) {
        // BEFORE updating window_end, check if string is valid and update pointers/set
        const newChar = string[windowEnd];
        
        while (charsInSubstring.has(newChar)) {
            const charAtStart = string[windowStart];
            charsInSubstring.delete(charAtStart);
            windowStart++;
        }
        
        const curValidSubstringLength = windowEnd - windowStart + 1;
        maxValidSubstringLength = Math.max(maxValidSubstringLength, curValidSubstringLength);
        
        charsInSubstring.add(newChar);
        windowEnd++;
    }
    
    return maxValidSubstringLength;
};

File: solutions/python/general-algos/strings/longest_substring_without_repeating_characters.py

class Solution:
    def lengthOfLongestSubstring(self, string: str) -> int:
        max_valid_substring_length = 0
        chars_in_substring = set()

        window_start = 0
        window_end = 0

        num_chars = len(string)
        while window_end < num_chars:
            # BEFORE updating window_end, check if string is valid and update pointers/set
            new_char = string[window_end]

            while new_char in chars_in_substring:
                char_at_start = string[window_start]
                chars_in_substring.remove(char_at_start)
                window_start += 1

            cur_valid_substring_length = window_end - window_start + 1
            max_valid_substring_length = max(max_valid_substring_length, cur_valid_substring_length)

            chars_in_substring.add(new_char)
            window_end += 1

        return max_valid_substring_length